server/service/schedule/schedule.go

```diff
@@ -0,0 +1,292 @@
+}
+
+// Start starts running the added jobs.
+//
+// All jobs must be added before calling Start.
+func (s *Schedule) Start() {
+	var m sync.Mutex // protects currentStart and currentWait.
+	currentStart := time.Now()
+	currentWait := 10 * time.Second
+
+	getWaitTimes := func() (start time.Time, wait time.Duration) {
+		m.Lock()
+		defer m.Unlock()
+
+		return currentStart, currentWait
+	}
+
+	setWaitTimes := func(start time.Time, wait time.Duration) {
+		m.Lock()
+		defer m.Unlock()
+
+		currentStart = start
+		currentWait = wait
+	}
+
+	if schedInterval := s.getSchedInterval(); schedInterval < currentWait {
+		setWaitTimes(currentStart, schedInterval)
+	}
+
+	var g sync.WaitGroup
+
+	schedTicker := time.NewTicker(currentWait)
+	g.Add(+1)
+	go func() {
+		defer g.Done()
+
+		for {
+			_, currWait := getWaitTimes()
+			level.Debug(s.logger).Log("msg", "waiting", "current wait time", currWait)
+
+			select {
+			case <-s.ctx.Done():
+				return
+
+			case <-schedTicker.C:
+				level.Debug(s.logger).Log("waiting", "done")
+
+				schedInterval := s.getSchedInterval()
+				schedTicker.Reset(schedInterval)
+
+				newStart := time.Now()
+				newWait := schedInterval
+				setWaitTimes(newStart, newWait)
+
+				if ok := s.acquireLock(); !ok {
+					continue
+				}
+
+				for _, job := range s.jobs {
+					level.Debug(s.logger).Log("msg", "starting", "jobID", job.ID)
+					if err := job.Fn(s.ctx); err != nil {
+						level.Error(s.logger).Log("jobID", job.ID, "err", err)
+						sentry.CaptureException(err)
+					}
+				}
+			}
+		}
+	}()
+
+	// Periodically check for config updates and resets the schedInterval for the previous loop.
+	configTicker := time.NewTicker(200 * time.Millisecond)
+	g.Add(+1)
+	go func() {
+		defer g.Done()
+
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			case <-configTicker.C:
+				level.Debug(s.logger).Log("msg", "config reload check")
+
+				configTicker.Reset(s.configReloadInterval)
+
+				schedInterval := s.getSchedInterval()
+				currStart, _ := getWaitTimes()
+
+				if s.configReloadIntervalFn == nil {
+					level.Debug(s.logger).Log("msg", "config reload interval method not set")
+					continue
+				}
+
+				newInterval, err := s.configReloadIntervalFn(s.ctx)
+				if err != nil {
+					level.Error(s.logger).Log("msg", "schedule interval config reload failed", "err", err)
+					sentry.CaptureException(err)
+					continue
+				}
+				if schedInterval == newInterval {
+					level.Debug(s.logger).Log("msg", "schedule interval unchanged")
+					continue
+				}
+				s.setSchedInterval(newInterval)
+
+				newWait := 10 * time.Millisecond
+				if time.Since(currStart) < newInterval {
+					newWait = newInterval - time.Since(currStart)
+				}
+				setWaitTimes(currStart, newWait)
+				schedTicker.Reset(newWait)
+
+				level.Debug(s.logger).Log("new schedule interval", newInterval, "new wait", newWait)
+			}
+		}
+	}()
+
+	go func() {
+		g.Wait()
+		level.Debug(s.logger).Log("msg", "done")
+		close(s.done)
```

Comment:
This was a bit confusing - I thought it was a bug because it closed `s.done` while the other goroutines check on `s.ctx.Done()`, but I think that's deliberate - this `s.done` is just to indicate that the scheduler has finished running, right? I think a comment explaining that would help.